### PR TITLE
fix(app-router): tag cached pages by route pattern

### PR DESCRIPTION
**Pull request by NathanDrake2406**
## What this changes
App Router page and route handler cache entries now carry implicit tags derived from the matched route pattern in addition to exact resolved pathname tags.

## Why
`revalidatePath(path, type)` targets `_N_T_` tags based on route files. Entries rendered at dynamic URLs need tags such as `_N_T_/blog/[slug]/layout` so layout-scoped invalidation reaches cached fetches beneath dynamic routes.

Refs #921

## Approach
Move implicit tag construction into `server/implicit-tags.ts` and wire generated RSC entries to pass `route.routeSegments` plus the leaf kind. The helper preserves exact pathname tags, root/index aliases, route groups, dynamic segments, page leaves, and route handler leaves.

This is scoped to App Router cache tag generation and does not change ISR storage semantics.

## Validation
- `vp test run tests/page-cache-tags.test.ts tests/isr-cache.test.ts tests/fetch-cache.test.ts -t "App Router page cache tags|revalidatePath|soft tags"`
- `vp test run tests/app-router.test.ts -t "generated code threads collected fetch tags into page ISR writes|generated code stores rscData in the ISR cache entry"`
- `vp test run tests/entry-templates.test.ts`
- `vp check`
- `vp run vinext#build`

### CODE REVIEW

File: packages/vinext/src/server/implicit-tags.ts [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+  appendUnique(tags, `${NEXT_CACHE_IMPLICIT_TAG_ID}/layout`);
+
+  if (!routePath.startsWith("/")) return;
+
+  const routeParts = routePath.split("/");
+  for (let i = 1; i < routeParts.length + 1; i++) {
+    let currentPathname = routeParts.slice(0, i).join("/");
+    if (!currentPathname) continue;
+
+    if (!currentPathname.endsWith("/page") && !currentPathname.endsWith("/route")) {
```
[the remark this answers is not in the document]
Comment (NathanDrake2406) [reply packages/vinext/src/server/implicit-tags.ts]: Fixed by deriving the leaf from its position in the route-file path instead of suffix matching the assembled pathname. Added regression coverage for literal `page` and `route` URL segments so intermediate directories still produce layout tags.

**Comment by NathanDrake2406** [comment]
/bigbonk review this PR. Focus on blocking correctness issues and test gaps.

**Comment by NathanDrake2406** [comment]
/bigbonk review this PR. Focus on blocking correctness issues and whether the tests cover the real behavior and meaningful edge cases.

File: tests/isr-cache.test.ts [hunk 2]
```diff
@@ -327,26 +328,13 @@ describe("triggerBackgroundRegeneration", () => {
...
-    tags.push(`_N_T_${built}/page`);
-    return tags;
+  function staticRouteSegments(pathname: string): string[] {
+    return pathname.split("/").filter(Boolean);
   }
 
   /** Helper: store a FETCH cache entry with path + implicit hierarchy tags. */
   async function seedEntry(path: string, body: string): Promise<void> {
-    const tags = [path, `_N_T_${path}`, ...deriveImplicitTags(path)];
+    const tags = buildPageCacheTags(path, [], staticRouteSegments(path), "page");
```
[the remark this answers is not in the document]
Comment (NathanDrake2406) [reply tests/isr-cache.test.ts]: Fixed by making the ISR test seeder accept route segments and adding dynamic route coverage: typed invalidation with `/blog/[slug]` hits an entry cached at `/blog/hello`, typed invalidation with `/blog/hello` does not, and no-type invalidation with `/blog/hello` still hits the exact pathname tag.

File: packages/vinext/src/server/implicit-tags.ts [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+function normalizeRouteSegment(segment: string): string | null {
+  if (!segment || segment === "." || segment.startsWith("@")) return null;
+
+  if (!segment.startsWith(":")) return segment;
+
+  const param = segment.slice(1);
+  if (param.endsWith("+")) return `[...${param.slice(0, -1)}]`;
+  if (param.endsWith("*")) return `[[...${param.slice(0, -1)}]]`;
+  return `[${param}]`;
```
[the remark this answers is not in the document]
Comment (NathanDrake2406) [reply packages/vinext/src/server/implicit-tags.ts]: Addressed by removing the unreachable colon-parameter conversion and documenting that app route segments are raw filesystem segments, with dynamic segments already in bracket notation. Kept the invisible segment guard because the helper can safely ignore root and slot-like segments if a future caller passes them.

File: tests/page-cache-tags.test.ts [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+      "/api/route/posts",
+      "_N_T_/api/route/posts",
+      "_N_T_/layout",
+      "_N_T_/api/layout",
+      "_N_T_/api/route/layout",
+      "_N_T_/api/route/posts/layout",
+      "_N_T_/api/route/posts/route",
+    ]);
+  });
+});
```
[the remark this answers is not in the document]
Comment (NathanDrake2406) [reply tests/page-cache-tags.test.ts:93]: Fixed by adding explicit coverage for catch-all and optional catch-all route segments so normalization keeps their bracket notation in derived layout/page tags.

File: packages/vinext/src/server/implicit-tags.ts [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+}
+
+function appendDerivedTags(tags: string[], routePath: string): void {
+  appendUnique(tags, `${NEXT_CACHE_IMPLICIT_TAG_ID}/layout`);
+
+  if (!routePath.startsWith("/")) return;
+
+  const routeParts = routePath.split("/");
+  const leafIndex = routeParts.length - 1;
+  for (let i = 1; i < routeParts.length + 1; i++) {
```
[the remark this answers is not in the document]
Comment (NathanDrake2406) [reply packages/vinext/src/server/implicit-tags.ts]: Addressed by changing the loop bound to `i <= routeParts.length`; behavior is unchanged.

File: packages/vinext/src/server/implicit-tags.ts [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+  const routeParts = routePath.split("/");
+  const leafIndex = routeParts.length - 1;
+  for (let i = 1; i <= routeParts.length; i++) {
+    let currentPathname = routeParts.slice(0, i).join("/");
+    if (!currentPathname) continue;
+
+    const isLeaf = i - 1 === leafIndex;
+    if (!isLeaf) {
+      currentPathname = `${currentPathname}${currentPathname.endsWith("/") ? "" : "/"}layout`;
```
[the remark this answers is not in the document]
Comment (james-elicx) [reply packages/vinext/src/server/implicit-tags.ts]: hmm, yeah, reading the logic, i dont think it's possible to end with `/` inside this loop because of the leaf check

File: packages/vinext/src/entries/app-rsc-entry.ts [hunk 7]
```diff
@@ -2178,7 +2161,9 @@ async function _handleRequest(request, __reqCtx, _mwCtx) {
     ) {
       const __cachedRouteResponse = await __readAppRouteHandlerCacheResponse({
         basePath: __basePath,
-        buildPageCacheTags: __pageCacheTags,
+        buildPageCacheTags: function(pathname, extraTags) {
```
[the remark this answers is not in the document]
Comment (NathanDrake2406) [reply packages/vinext/src/entries/app-rsc-entry.ts]: Fixed by extracting the route-handler cache-tag callback once inside `_handleRequest` and passing that helper to both route handler execution paths. Updated the generated-entry assertions and snapshots.

File: packages/vinext/src/server/implicit-tags.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (NathanDrake2406) [reply packages/vinext/src/server/implicit-tags.ts]: Fixed by dropping the unreachable trailing-slash guard and appending `/layout` directly. The route cache path is generated without trailing slashes, so behavior is unchanged.
